### PR TITLE
Adds the `qml.gradients` module to the documentation

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The `qml.gradients` module is not currently documented.

**Description of the Change:**

- Adds the `qml.gradients` module to the documentation
- Updates docstrings to take into account recent changes

**Benefits:** n/a

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
[ch8556]

File: pennylane/gradients/__init__.py [hunk 1]
```diff
@@ -11,7 +11,266 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""This subpackage contains quantum gradient transforms."""
+"""
+
+Quantum gradient transforms are strategies of computing the gradient of a quantum
+circuit by **transforming** the quantum circuit into one or more gradient circuits.
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:17]:
```suggestion
Quantum gradient transforms are strategies for computing the gradient of a quantum
circuit that work by **transforming** the quantum circuit into one or more gradient circuits.
```

File: pennylane/gradients/__init__.py [hunk 2]
```diff
@@ -11,7 +11,266 @@
...
+    generate_shifted_tapes
+    compute_vjp
+    batch_vjp
+    vjp
+
+
+Registering autodifferentiation gradients
+-----------------------------------------
+
+All PennyLane QNode's are automatically differentiable, and can be included
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:74]:
```suggestion
All PennyLane QNodes are automatically differentiable, and can be included
```

File: pennylane/gradients/__init__.py [hunk 3]
```diff
@@ -11,7 +11,266 @@
...
+seamlessly within an autodiff pipeline. When creating a :class:`~.QNode`, the
+strategy for determining the optimal differentiation strategy is *automated*,
+and takes into account the circuit, device, autodiff framework, and metadata
+(such as whether finite shots are used).
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:78]:
```suggestion
(such as whether a finite number of shots are used).
```

File: pennylane/gradients/__init__.py [hunk 4]
```diff
@@ -11,7 +11,266 @@
...
+
+.. code-block:: python
+
+    @qml.qnode(dev)
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:82]: Perhaps use a more informative code block here that specifies an interface, shots, etc. to match with the previous description.

File: pennylane/gradients/__init__.py [hunk 5]
```diff
@@ -11,7 +11,266 @@
...
+    def circuit(weights):
+        ...
+
+In particular:
+
+- When using a simulator device with exact measurement statistics, backpropagation
+  is preferred due to performance and memory improvements.
+
+- When using a hardware device, or a simulator with a finite-number of shots,
+  a quantum gradient transform---such as the parameter-shift rule---are preferred.
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:92]:
```suggestion
  a quantum gradient transform---such as the parameter-shift rule---is preferred.
```

File: pennylane/gradients/__init__.py [hunk 6]
```diff
@@ -11,7 +11,266 @@
...
+    @qml.qnode(dev)
+    def circuit(weights):
+        ...
+
+In particular:
+
+- When using a simulator device with exact measurement statistics, backpropagation
+  is preferred due to performance and memory improvements.
+
+- When using a hardware device, or a simulator with a finite-number of shots,
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:91]:
```suggestion
- When using a hardware device, or a simulator with a finite number of shots,
```

File: pennylane/gradients/__init__.py [hunk 7]
```diff
@@ -11,7 +11,266 @@
...
+
+>>> decorated_circuit(weights)
+tensor([[-0.04673668, -0.09442394, -0.14409127],
+        [ 0.04673668,  0.09442394,  0.14409127]], requires_grad=True)
+
+.. note::
+
+    If your circuit contains any operations not supported by the gradient
+    transform, the transform will attempt to automatically decompose the
+    circuit to find one supporting gradients.
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:170]:
```suggestion
    circuit into only operations that support gradients.
```

File: pennylane/gradients/__init__.py [hunk 8]
```diff
@@ -11,7 +11,266 @@
...
+    with qml.tape.JacobianTape() as tape:
+        qml.RX(weights[0], wires=0)
+        qml.RY(weights[1], wires=1)
+        qml.CNOT(wires=[0, 1])
+        qml.RX(weights[2], wires=1)
+        qml.expval(qml.PauliZ(1))
+
+Unlike when transforming a QNode, transforming a tape directly
+will perform no implicit quantum device evaluation. Instead, it returns
+the processed tapes, and post-processing function, which together
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:230]:
```suggestion
the processed tapes, and a post-processing function, which together
```

File: pennylane/gradients/__init__.py [hunk 9]
```diff
@@ -11,7 +11,266 @@
...
+This can be useful if the underlying circuits representing the gradient
+computation need to be analyzed.
+
+The output tapes can then be evaluated and post-processed to retrieve
+the gradient:
+
+>>> dev = qml.device("default.qubit", wires=2)
+>>> from pennylane.interfaces.batch import execute
+>>> fn(execute(gradient_tapes, dev, None))
+[[-0.09347337 -0.18884787 -0.28818254]]
```
Comment (glassnotes) [inline pennylane/gradients/__init__.py:251]: Great example here!

File: pennylane/gradients/finite_difference.py [hunk 10]
```diff
@@ -183,11 +183,11 @@ def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
 
 @gradient_transform
 def finite_diff(tape, argnum=None, h=1e-7, approx_order=1, n=1, strategy="forward", f0=None):
-    r"""Generate the finite-difference tapes and postprocessing methods required
-    to compute the gradient of a gate parameter with respect to its outputs.
+    r"""Transform a QNode to compute the finite-difference gradient of all gate
+    parameters with respect to its outputs.
```
Comment (glassnotes) [inline pennylane/gradients/finite_difference.py:187]:
```suggestion
    parameters with respect to its inputs.
```
?

File: pennylane/gradients/finite_difference.py [hunk 11]
```diff
@@ -183,11 +183,11 @@ def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
...
-    to compute the gradient of a gate parameter with respect to its outputs.
+    r"""Transform a QNode to compute the finite-difference gradient of all gate
+    parameters with respect to its outputs.
 
     Args:
-        tape (.QuantumTape): quantum tape to differentiate
+        qnode (.QNode or .QuantumTape): quantum tape or QNode to differentiate
         argnum (int or list[int] or None): Trainable parameter indices to differentiate
             with respect to. If not provided, the derivatives with respect to all
             trainable indices are returned.
```
Comment (glassnotes) [inline pennylane/gradients/finite_difference.py:193]:
```suggestion
            trainable parameters are returned.
```

File: pennylane/gradients/gradient_transform.py [hunk 12]
```diff
@@ -87,7 +87,7 @@ class gradient_transform(qml.batch_transform):
 
     .. code-block:: python
 
-        @quantum_gradient
+        @gradient_transform
```
Comment (glassnotes) [inline pennylane/gradients/gradient_transform.py:90]: I cannot seem to leave a suggestion, but on line 59 above, change outputs -> output.

File: pennylane/gradients/parameter_shift.py [hunk 13]
```diff
@@ -354,12 +354,11 @@ def processing_fn(results):
 def param_shift(
     tape, argnum=None, shift=np.pi / 2, gradient_recipes=None, fallback_fn=finite_diff, f0=None
 ):
-    r"""Generate the parameter-shift tapes and postprocessing methods required
-    to compute the gradient of a gate parameter with respect to an
-    expectation value.
+    r"""Transform a QNode to compute the parameter-shift gradient of all gate
+    parameters with respect to its outputs.
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift.py:358]: Same comment as earlier, outputs -> inputs?

File: pennylane/gradients/parameter_shift.py [hunk 14]
```diff
@@ -443,19 +448,66 @@ def param_shift(
...
-     [ 0.69916862  0.34072424  0.69202359]]
+    ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+    >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+    >>> qml.jacobian(circuit)(params)
+    tensor([[-0.38751725, -0.18884792, -0.38355708],
+            [ 0.69916868,  0.34072432,  0.69202365]], requires_grad=True)
+
+    .. UsageDetails::
+
+        This gradient transform can also be applied directly to :class:`~.QNode` objects:
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift.py:468]:
```suggestion
        This gradient transform can be applied directly to :class:`~.QNode` objects:
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 15]
```diff
@@ -472,8 +472,8 @@ def param_shift_cv(
     f0=None,
     force_order2=False,
 ):
-    r"""Generate the CV parameter-shift tapes and postprocessing methods required
-    to compute the gradient of a gate parameter with respect to the CV output.
+    r"""Transform a continuous-variable QNode to compute the parameter-shift gradient of all gate
+    parameters with respect to its outputs.
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:476]: Outputs -> inputs?

File: pennylane/gradients/parameter_shift_cv.py [hunk 16]
```diff
@@ -545,47 +551,88 @@ def param_shift_cv(
...
-    >>> tape.trainable_params = {0, 2}
-    >>> gradient_tapes, fn = qml.gradients.param_shift_cv(tape, dev)
-    >>> res = dev.batch_execute(gradient_tapes)
-    >>> fn(res)
-    array([[-0.32487113, -0.87049853]])
+    This transform can be registered directly as the quantum gradient transform
+    to use during autodifferentiation:
+
+    >>> dev = qml.device("default.gaussian", wires=2)
+    >>> @qml.qnode(dev)#, gradient_fn=qml.gradients.param_shift_cv)
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:587]:
```suggestion
    >>> @qml.qnode(dev, gradient_fn=qml.gradients.param_shift_cv)
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 17]
```diff
@@ -545,47 +551,88 @@ def param_shift_cv(
...
+    ...     qml.Squeezing(params[0], params[1], wires=[0])
+    ...     qml.Squeezing(params[2], params[3], wires=[0])
+    ...     return qml.expval(qml.NumberOperator(0))
+    >>> params = np.array([0.1, 0.2, 0.3, 0.4], requires_grad=True)
+    >>> qml.jacobian(circuit)(params)
+    array([ 0.87516064,  0.01273285,  0.88334834, -0.01273285])
+
+    .. UsageDetails::
+
+        This gradient transform can also be applied directly to :class:`~.QNode` objects:
```
Comment (glassnotes) [inline pennylane/gradients/parameter_shift_cv.py:598]:
```suggestion
        This gradient transform can be applied directly to :class:`~.QNode` objects:
```

**Review by glassnotes — APPROVED** [review]
@josh146 great to finally see this come together! Basically just typos and a few places where I was confused about the phrasing. Otherwise this is good to go :tada:

File: pennylane/gradients/parameter_shift.py (hunk 13 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift.py:358]:
```suggestion
    parameters with respect to its inputs.
```

File: pennylane/gradients/parameter_shift_cv.py (hunk 15 above)
Comment (josh146) [reply pennylane/gradients/parameter_shift_cv.py:476]:
```suggestion
    parameters with respect to its inputs.
```

File: pennylane/gradients/__init__.py [hunk 18]
```diff
@@ -11,7 +11,266 @@
...
+
+This can be useful if the underlying circuits representing the gradient
+computation need to be analyzed.
+
+The output tapes can then be evaluated and post-processed to retrieve
+the gradient:
+
+>>> dev = qml.device("default.qubit", wires=2)
+>>> from pennylane.interfaces.batch import execute
+>>> fn(execute(gradient_tapes, dev, None))
```
Comment (anthayes92) [inline pennylane/gradients/__init__.py:250]: Would it be worth including a little detail  above on how to use the post processing function? E.g what type of arg it expects.

**Review by anthayes92 — APPROVED** [review]

File: pennylane/gradients/__init__.py (hunk 4 above)
Comment (josh146) [reply pennylane/gradients/__init__.py:82]: done!

File: pennylane/gradients/gradient_transform.py (hunk 12 above)
Comment (josh146) [reply pennylane/gradients/gradient_transform.py:90]: done!

File: pennylane/gradients/__init__.py (hunk 18 above)
Comment (josh146) [reply pennylane/gradients/__init__.py:250]: Good suggestion, thanks @anthayes92!
